### PR TITLE
4777 - Implement IDS Mask

### PR DESCRIPTION
**Pull request by EdwardCoyle**
**Explain the details for making this change. What existing problem does the pull request solve?**
This PR implements the input masking system from the Enterprise components as a Mixin combined with IdsInput.  This allows for controlled entry into input fields, throwing away any unexpected/unwanted input.  Major features include:
- simple pattern-based content control (array-based) for string formatting as the user types
- mask functions (dynamic masks)
- supports number formatting as the user types
- supports date formatting as the user types
- can display a visual guide of the mask (pattern-based masks only)

Note that some localization-specific features aren't currently available (for example, formatting a number to contain thousands separators isn't possible until we bring over the Locale system, since placement of the thousands-separator in a number is different depending on the "groups" defined in Locale.

**Related github/jira issue (required)**:
Closes infor-design/enterprise#4777

**Steps necessary to review your pull request (required)**:
- Pull, build, run
- Open http://localhost:4300/ids-mask
- Try all the mask fields.  Any input that was possible in 4.x should be possible in these fields.
- Try mask functions (prefix/suffix) - http://localhost:4300/ids-mask/prefix-suffix
- Try guide configurations (pattern mask only) - http://localhost:4300/ids-mask/guides
- Try "string-ified" mask definition by attribute (pattern mask only) - http://localhost:4300/ids-mask/mask-attribute
- Try leading zeroes in the number mask - http://localhost:4300/ids-mask/number-leading-zeros

More configurations, custom masks, etc are demonstrated in the tests.

**Included in this Pull Request**:
- [x] An e2e or functional test for the bug or feature.
- [x] A note to the change log.

### CODE REVIEW

**Review by tmcconechy — DISMISSED** [review]
Looks good. Some small comments..

Good for now, but wonder if we will need some way to attach it to the field easier. Maybe we could parse the placeholder somehow or allow a "simple" mask as a string prop on the input. We can take feedback on this tho.

**Review by deep7102 — APPROVED** [review]
Looks good 🍌 
nothing else found follow as Tim's comment.

File: src/ids-mask/ids-mask-mixin.js [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+  props.MASK_RETAIN_POSITIONS,
+  props.MASK_OPTIONS,
+];
+
+/**
+ * Adds validation to any input field
+ * @param {any} superclass Accepts a superclass and creates a new subclass from it
+ * @returns {any} The extended object
+ */
+// @ts-ignore
```
[the remark this answers is not in the document]
Comment (EdwardCoyle) [reply src/ids-mask/ids-mask-mixin.js:21]: I had pulled his changes into this PR and if you remove this ignore statement, there's still an error on this line.  Seems to work regardless.

File: src/ids-input/ids-input.js [hunk 2]
```diff
@@ -19,12 +19,37 @@ import { IdsEventsMixin } from '../ids-base/ids-events-mixin';
 import { IdsKeyboardMixin } from '../ids-base/ids-keyboard-mixin';
 import { IdsClearableMixin } from '../ids-base/ids-clearable-mixin';
 import { IdsDirtyTrackerMixin } from '../ids-base/ids-dirty-tracker-mixin';
+import IdsMaskMixin from '../ids-mask/ids-mask-mixin';
```
[the remark this answers is not in the document]
Comment (EdwardCoyle) [reply src/ids-input/ids-input.js:22]: `IdsMaskMixin` is the only thing exported from that file, so I figured making it a default export made sense.  I can make it "look" like the other mixins, but functionally I thought this made more sense?

File: src/ids-mask/README.md [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# Ids Mask API
+
```
[the remark this answers is not in the document]
Comment (EdwardCoyle) [reply src/ids-mask/README.md:2]: Yeahhhhh that's a mistake.  I'll move it 👍

File: src/ids-input/ids-input.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmcconechy) [reply src/ids-input/ids-input.js:22]: I think maybe we could make them all look a like but was thinking of exporting these all from IdsElement instead so its not that important. Maybe do it if its very simple for you

File: src/ids-mask/ids-mask-mixin.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tmcconechy) [reply src/ids-mask/ids-mask-mixin.js:21]: ok i thought we didnt need any of these now. Maybe you have to turn off the editor TS linting for JS files ( i had to)

[the remark this answers is not in the document]
Comment (EdwardCoyle) [reply src/ids-mask/ids-mask-mixin.js:21]: Hmm, I'd forgotten that there's an editor level setting for this.  Just did this and it's fixed

File: src/ids-input/ids-input.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (EdwardCoyle) [reply src/ids-input/ids-input.js:22]: I can change it but it's technically not correct. Using the bracket syntax is saying "there are other things in this file that can be accessed through imports" which isn't correct in my case.

Actually I wouldn't mind taking on a separate issue trying to standardize the imports/exports throughout, if that's a task we want to do.  Exporting from `IdsElement` isn't a bad idea, but the source code files for the Mixins themselves should be exporting a default class/method/etc.  It's something we're not totally doing right atm.

[the remark this answers is not in the document]
Comment (tmcconechy) [reply src/ids-input/ids-input.js:22]: Ok sounds good. lets hit it later then....

**Review by tmcconechy — APPROVED** [review]

**Review by deep7102 — APPROVED** [review]
🍌

**Pull request merged**